build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/ParseExternalGitDependencies.cs

```diff
@@ -0,0 +1,56 @@
+﻿
+using System.Collections.Generic;
+using System.IO;
+using System.Text.RegularExpressions;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.Android.BuildTools.PrepTasks
+{
+	public class ParseExternalGitDependencies : Task
+	{
+		[Required]
+		public string ExternalFilePath { get; set; }
+		
+		/* %(ExternalGitDependencies.Owner)     - Repo owner
+		 * %(ExternalGitDependencies.Name)      - Repo name
+		 * %(ExternalGitDependencies.Branch)    - Branch name
+		 * %(ExternalGitDependencies.Identity)  - Commit hash
+		 */
+		[Output]
+		public ITaskItem[] ExternalGitDependencies { get; set; }
+
+		static readonly Regex externalRegex = new Regex (
+			@"^(?<comment>\#\s*)?(?<owner>.*)\/(?<repo>.*):(?<branch>.*)@(?<commit>.*)$",
+			RegexOptions.Compiled);
+
+		public override bool Execute ()
+		{
+			if (!File.Exists (ExternalFilePath)) {
+				Log.LogError($"Unable to find dependency file at: {ExternalFilePath}");
+				return false;
+			}
+
+			string[] unparsedExternals = File.ReadAllLines (ExternalFilePath);
+			var externals = new List<TaskItem> (unparsedExternals.Length);
+
+			foreach (string external in unparsedExternals) {
+				Match match = externalRegex.Match (external);
+				if (match != null && match.Success) {
+					if (match.Groups["comment"].Success) {
+						// Ignore matching lines which start with '#'.
+						continue;
+					}
+					var e = new TaskItem (match.Groups["commit"].Value);
+					e.SetMetadata ("Owner", match.Groups["owner"].Value);
+					e.SetMetadata ("Name", match.Groups["repo"].Value);
+					e.SetMetadata ("Branch", match.Groups["branch"].Value);
+					externals.Add (e);
```

Comment:
Philosophical question: The `.external` file spec has an **Advanced** section which states that some things can be overridden via environment variables:

> A build can optionally override any of the declared externals by setting an environment variable or build property with the format: `organization/repository`=`branch@commit`. The specified value is used as the `branch@commit` instead of the declared one for a matching `organization/repository` line in the `.external` file.

One Fatal Flaw with this approach is that it *only* works on Windows:

```cmd
> set FOO/BAR=baz@qux
> echo %FOO/BAR%
baz@qux
```

Do we want to even *consider* this "advanced" functionality?  Implementation-wise, I don't think it would be difficult to support, but...it can't possibly work on Unix.

Reply:
I'm not too concerned with supporting this use case at the moment. I'm struggling to picture a common scenario where this workflow would be more useful _and_ easier than updating the value in the .external file. I think this can be revisited in the future if we find a need for it.